File: 2026/04-April/03-Friday/solution.java
import java.util.*;

class Solution {
    private Integer[][] dp;
    private int[][] robotData;
    private int[] walls;
    private int n;

    public int maxWalls(int[] robots, int[] distance, int[] walls) {
        n = robots.length;

        robotData = new int[n][2];
        for (int i = 0; i < n; i++) {
            robotData[i][0] = robots[i];
            robotData[i][1] = distance[i];
        }

        Arrays.sort(robotData, Comparator.comparingInt(a -> a[0]));
        Arrays.sort(walls);
        this.walls = walls;

        dp = new Integer[n][2];

        return dfs(n - 1, 1);
    }

    private int dfs(int i, int nextRight) {
        if (i < 0) return 0;

        if (dp[i][nextRight] != null) return dp[i][nextRight];

        int pos = robotData[i][0];
        int dist = robotData[i][1];

        // Left
        int leftBoundary = pos - dist;
        if (i > 0) {
            leftBoundary = Math.max(leftBoundary, robotData[i - 1][0] + 1);
        }

        int l = lowerBound(walls, leftBoundary);
        int r = lowerBound(walls, pos + 1);
        int resLeft = dfs(i - 1, 0) + (r - l);

        // Right
        int rightBoundary = pos + dist;
        if (i + 1 < n) {
            int nextPos = robotData[i + 1][0];
            int nextDist = robotData[i + 1][1];

            if (nextRight == 0) {
                rightBoundary = Math.min(rightBoundary, nextPos - nextDist - 1);
            } else {
                rightBoundary = Math.min(rightBoundary, nextPos - 1);
            }
        }

        int resRight;
        if (rightBoundary < pos) {
            resRight = dfs(i - 1, 1);
        } else {
            l = lowerBound(walls, pos);
            r = lowerBound(walls, rightBoundary + 1);
            resRight = dfs(i - 1, 1) + (r - l);
        }

        return dp[i][nextRight] = Math.max(resLeft, resRight);
    }

    private int lowerBound(int[] arr, int target) {
        int l = 0, r = arr.length;
        while (l < r) {
            int m = (l + r) / 2;
            if (arr[m] < target) l = m + 1;
            else r = m;
        }
        return l;
    }
}

File: 2026/04-April/03-Friday/solution.kt
//updated
import java.util.*

class Solution {
    private lateinit var dp: Array<IntArray?>
    private lateinit var robotData: Array<IntArray>
    private lateinit var walls: IntArray
    private var n = 0

    fun maxWalls(robots: IntArray, distance: IntArray, walls: IntArray): Int {
        n = robots.size

        robotData = Array(n) { IntArray(2) }
        for (i in 0 until n) {
            robotData[i][0] = robots[i]
            robotData[i][1] = distance[i]
        }

        robotData.sortBy { it[0] }
        walls.sort()
        this.walls = walls

        dp = Array(n) { IntArray(2) { -1 } }

        return dfs(n - 1, 1)
    }

    private fun dfs(i: Int, nextRight: Int): Int {
        if (i < 0) return 0

        if (dp[i][nextRight] != -1) return dp[i][nextRight]

        val pos = robotData[i][0]
        val dist = robotData[i][1]

        var leftBoundary = pos - dist
        if (i > 0) {
            leftBoundary = maxOf(leftBoundary, robotData[i - 1][0] + 1)
        }

        var l = lowerBound(walls, leftBoundary)
        var r = lowerBound(walls, pos + 1)
        val resLeft = dfs(i - 1, 0) + (r - l)

        var rightBoundary = pos + dist
        if (i + 1 < n) {
            val nextPos = robotData[i + 1][0]
            val nextDist = robotData[i + 1][1]

            rightBoundary = if (nextRight == 0) {
                minOf(rightBoundary, nextPos - nextDist - 1)
            } else {
                minOf(rightBoundary, nextPos - 1)
            }
        }

        val resRight = if (rightBoundary < pos) {
            dfs(i - 1, 1)
        } else {
            l = lowerBound(walls, pos)
            r = lowerBound(walls, rightBoundary + 1)
            dfs(i - 1, 1) + (r - l)
        }

        dp[i][nextRight] = resLeft.coerceAtLeast(resRight)
        return dp[i][nextRight]
    }

    private fun lowerBound(arr: IntArray, target: Int): Int {
        var l = 0
        var r = arr.size
        while (l < r) {
            val m = (l + r) / 2
            if (arr[m] < target) l = m + 1 else r = m
        }
        return l
    }
}
